src/temporal_terms.jl

```diff
@@ -0,0 +1,56 @@
+# Basic Lag:
+
+# syntax:
+"""
+    lag(term, nsteps)
+This `@formula` term is used to introduce lagged variables.
+For example `lag(x,1)` effectively adds a new column containing
+the value of the `x` column from the previous row.
+If there are no such row (e.g. because this is the first row), then the lagged column will
+contain `missing`.
+
+Note: this is only a basic row-wise lag operation.
+It is up to the user to ensure that that there data is sorted by the temporal variable,
+and that their observations are spaced with regular time-steps.
+(Which may require adding extra-rows filled with `missing` values.)
+"""
+lag(term, nsteps) = error("`lag`  should only be used within a @formula");
+
+# type of model where syntax applies:
+const LAG_CONTEXT = Any
+
+# struct for behavior
+struct LagTerm{T<:AbstractTerm} <: AbstractTerm
+    term::T
+    nsteps::Int
+end
+
+Base.show(io::IO, lag::LagTerm) = print(io, "lag($(lag.term), $(lag.nsteps))")
+
+function apply_schema(t::FunctionTerm{typeof(lag)}, sch, mod::Type{<:LAG_CONTEXT})
+    @assert length(t.args_parsed) == 2
+    term_parsed, nsteps_parsed = t.args_parsed
+
+    term = apply_schema(term_parsed, sch, mod)
+
+    (nsteps_parsed isa ConstantTerm) ||
+        throw(ArgumentError("Lag step must be a number (got $step_parsed)"))
+
+    nsteps = nsteps_parsed.n
+    return LagTerm(term, nsteps)
+end
+
+function modelcols(lag::LagTerm, d::NamedTuple)
```

Comment:
this will only work when `d` is a "column table" (e.g. named tuple of vectors), not for single-row namedtuples.  so better to specialize on `Tables.ColumnTable`, and throw an error that says that lag is only supported for in-memory data.

Reply:
Docs for extending need updating

Reply:
It would be nice to support those for row tables.
I guess via a buffer in the term,
That gets iterated each time a row is processed.
But needs to be reset whenever one starts again.

Where can I find a example of an opt that works on row tables?

Reply:
There isn't one currently (at least not one like this, where the behavior is going to be different). @nalimilan and I talked a bit about using the idea of "stateful transforms" from patsy (the formula library behind python statsmodels): https://patsy.readthedocs.io/en/latest/stateful-transforms.html

But haven't looked into that in any depth.

Reply:
I feel like it might be not too crazy
to have
`modelrows(lag::LagTerm, lag_term_state, row::NamedTuple, row_iterator_start)`

Then this could be implemeted witha `CircularBuffer`

Reply:
Maybe better start with support for column tables only, and discuss row tables separately. That way we can focus on the design here.

Reply:
Indeed